File: python/tracing/dify/02_chat_streaming.py
#!/usr/bin/env python3
"""Streaming Dify chat message traced after stream consumption."""

from _shared import DifyExampleRuntime, collect_stream_answer, print_result


def main() -> None:
    workflow_name = "dify_chat_streaming.workflow"
    with DifyExampleRuntime(workflow_name) as runtime:
        client = runtime.chat_client()
        response = client.create_chat_message(
            inputs={},
            query="Stream a short Dify response.",
            user=runtime.user("stream"),
            response_mode="streaming",
        )
        response.raise_for_status()
        answer = collect_stream_answer(response)
        summary = {"workflow": workflow_name, "answer": answer}
        runtime.set_result(summary)
        print_result("chat_streaming", summary)


if __name__ == "__main__":
    main()

File: typescript/tracing/dify/02_chat_streaming.ts
#!/usr/bin/env node
import { ChatClient } from "dify-client";
import { withDifyRuntime } from "./_shared.js";

const workflowName = "dify_typescript_chat_streaming.workflow";

await withDifyRuntime(workflowName, async (runtime) => {
  const client = new ChatClient({
    apiKey: runtime.key("DIFY_CHAT_API_KEY"),
    baseUrl: runtime.baseUrl,
  });
  const stream = await client.createChatMessage({
    inputs: {},
    query: "Stream a short Dify response.",
    user: "respan-dify-ts-stream",
    response_mode: "streaming",
  });
  if (!(Symbol.asyncIterator in stream)) throw new Error("Expected a Dify stream");

  const events: string[] = [];
  let answer = "";
  for await (const event of stream) {
    const data = event.data && typeof event.data === "object" ? event.data : {};
    if (typeof data.event === "string") events.push(data.event);
    if (typeof data.answer === "string") answer += data.answer;
  }
  const result = { answer, events, status: stream.status };
  runtime.setResult(result);
  return result;
});
